model/riscv_insts_zaamo.sail

```diff
@@ -34,33 +44,57 @@ mapping encdec_amoop : amoop <-> bits(5) = {
   AMOMIN  <-> 0b10000,
   AMOMAX  <-> 0b10100,
   AMOMINU <-> 0b11000,
-  AMOMAXU <-> 0b11100
+  AMOMAXU <-> 0b11100,
+  AMOCAS  <-> 0b00101,
 }
 
 mapping clause encdec = AMO(op, aq, rl, rs2, rs1, size, rd)
-  <-> encdec_amoop(op) @ bool_bits(aq) @ bool_bits(rl) @ encdec_reg(rs2) @ encdec_reg(rs1) @ 0b0 @ size_enc(size) @ encdec_reg(rd) @ 0b0101111
-  when currentlyEnabled(Ext_Zaamo) & amo_width_valid(size)
+  <-> encdec_amoop(op) @ bool_bits(aq) @ bool_bits(rl) @ encdec_reg(rs2) @ encdec_reg(rs1) @ size_enc_wide(size) @ encdec_reg(rd) @ 0b0101111
+  when amo_encoding_valid(size, op, rs2, rd)
+
+// AMOCAS.D for RV32 and AMOCAS.Q for RV64 operate on a pair of
+// register. When the first register of a source register pair is
+// x0, then both halves of the pair read as zero.
+
+val rX_amo_bits : forall 'w, 'w in {1, 2, 4, 8, 16} & 'w <= xlen_bytes * 2 . (regidx, int('w)) -> bits('w * 8)
+function rX_amo_bits(i, width_bytes) =
+  if i != zreg then {
+    if width_bytes <= xlen_bytes then X(i)[width_bytes * 8 - 1 .. 0] else X(i + 1) @ X(i)
+  } else zeros()
+
+val wX_amo_bits : forall 'w, 'w in {1, 2, 4, 8, 16} & 'w <= xlen_bytes * 2 . (regidx, int('w), bits('w * 8)) -> unit
+function wX_amo_bits(i, width_bytes, data) =
+  if i != zreg then {
+    if width_bytes <= xlen_bytes then X(i) = sign_extend(data) else {
+      X(i) = data[xlen - 1 .. 0];
+      X(i + 1) = data[xlen * 2 - 1 .. xlen];
+    }
+  }
+
+// X_amo allows accessing parts of registers (writes are sign-extended),
+// or register pairs for AMOCAS.
+overload X_amo = {rX_amo_bits, wX_amo_bits}
```

Comment:
Is this the same behaviour as for Zilsd? In that case I'd suggest using something like X_pair() and moving it into a more central file.

Reply:
Yeah it is. Good idea.